### PR TITLE
[Streaming Telemetry 1] Prepares TopicPartitionChannel for telemetry 

### PR DESCRIPTION
**Pull request by sfc-gh-rcheng**
- Requires TelemetryService not null in TopicPartitionChannel
- Adds offset setters in TopicPartitionChannel to update offset and SnowflakeTelemetryChannelStatus offsets at the same time
- Moves JMX into new SnowflakeTelemetryChannelStatus 

Changes coming in [part 2 PR](https://github.com/snowflakedb/snowflake-kafka-connector/pull/694)
- SnowflakeTelemetryChannelStatus becomes a telemetry object
- Enables reportKafkaPartitionStart/Usage telemetry

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 1]
```diff
@@ -284,12 +291,6 @@ public TopicPartitionChannel(
               + " correct offset instead",
           this.getChannelName());
     }
-
-    if (enableCustomJMXMonitoring) {
```
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:288]: moved into snowflakeTelemetryChannelStatus constructor

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 2]
```diff
@@ -1095,10 +1127,9 @@ protected SnowflakeTelemetryService getTelemetryServiceV2() {
     return this.telemetryServiceV2;
   }
 
-  protected void setLatestConsumerOffset(long consumerOffset) {
```
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1098]: moved up to the other offset setters

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+import static com.snowflake.kafka.connector.internal.metrics.MetricsUtil.constructMetricName;
+
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.MetricRegistry;
+import com.snowflake.kafka.connector.internal.KCLogger;
+import com.snowflake.kafka.connector.internal.metrics.MetricsJmxReporter;
+import com.snowflake.kafka.connector.internal.metrics.MetricsUtil;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class SnowflakeTelemetryChannelStatus {
```
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:29]: Will extend SnowflakeTelemetryBasicInfo and become a proper telemetry object in the next PR

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryServiceV2.java
Comment (sfc-gh-rcheng) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryServiceV2.java:1]: moved into streaming/telemetry folder

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java [hunk 4]
```diff
@@ -817,27 +835,52 @@ public void testBigAvroBufferBytesThreshold() throws Exception {
   }
 
   @Test
-  public void testRegisterJmxMetrics() {
```
Comment (sfc-gh-rcheng) [inline src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java:820]: moved into SnowflakeTelemetryChannelStatusTest

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 5]
```diff
@@ -128,6 +131,13 @@ public SnowflakeSinkServiceV2(
     this.partitionsToChannel = new HashMap<>();
 
     this.tableName2SchemaEvolutionPermission = new HashMap<>();
+
+    // jmx
+    String connectorName =
+        conn == null || Strings.isNullOrEmpty(this.conn.getConnectorName())
+            ? "default_connector"
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:138]: Ideally this is not possible since name is a requirement but good to check.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 6]
```diff
@@ -1065,6 +1068,24 @@ public String getChannelName() {
...
+    this.snowflakeTelemetryChannelStatus.setOffsetPersistedInSnowflake(offsetPersistedInSnowflake);
+  }
+
+  private void setProcessedOffset(long processedOffset) {
+    this.processedOffset.set(processedOffset);
+    this.snowflakeTelemetryChannelStatus.setProcessedOffset(processedOffset);
+  }
+
+  protected void setLatestConsumerOffset(long consumerOffset) {
+    if (consumerOffset > this.latestConsumerOffset) {
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1083]: we arent checking this in original code, so lets not add this

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java (hunk 3 above)
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:29]: nit: javadocs fine if coming in next PR

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+      new KCLogger(SnowflakeTelemetryChannelStatus.class.toString());
+
+  public static final long NUM_METRICS = 3; // update when new metrics are added
+  private static final long INITIAL_OFFSET = -1;
+
+  // channel properties
+  private final String channelName;
+  private final MetricsJmxReporter metricsJmxReporter;
+
+  private final AtomicLong startTime;
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:40]: this seems unused, remove if not needed

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+  // offsets
+  private AtomicLong offsetPersistedInSnowflake;
+  private AtomicLong processedOffset;
+  private long latestConsumerOffset;
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:45]: do we need atomic long?

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java [hunk 9]
```diff
@@ -817,27 +835,52 @@ public void testBigAvroBufferBytesThreshold() throws Exception {
...
-    MetricRegistry metricRegistry = Mockito.spy(MetricRegistry.class);
-    MetricsJmxReporter metricsJmxReporter =
-        Mockito.spy(new MetricsJmxReporter(metricRegistry, TEST_CONNECTOR_NAME));
+  public void testTopicPartitionChannelInvalidJmxReporter() throws Exception {
```
Comment (sfc-gh-japatel) [inline src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java:838]: lets also add a test where there are >1 channels and we verify the metrics values after inserting few records in each channel.

Comment (sfc-gh-japatel) [reply src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java:838]: check out `SnowflakeTelemetryPipeStatusMetricsIT`

**Review by sfc-gh-japatel — APPROVED** [review]
overall lgtm, couple of questions and one request to add a new test.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java (hunk 8 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:45]: SnowflakeTelemetryPipeStatus uses AtomicLongs, so I figured this should follow precedent

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java (hunk 3 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:29]: Added javadocs

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java (hunk 9 above)
Comment (sfc-gh-rcheng) [reply src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java:838]: thanks for the suggestion, there's a similar test in the next PR but it only has one channel. 

Added metrics verification to testStreamingIngest_multipleChannelPartitions_withMetrics in SnowflakeSinkServiceV2IT.java

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java (hunk 7 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java]: removed

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java (hunk 5 above)
Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:138]: My two cents: we should not do this if the assumption is connector name should never be null, you can add an assert instead

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 10]
```diff
@@ -1036,6 +1037,8 @@ private SnowflakeStreamingIngestChannel openChannelForTable() {
   public void closeChannel() {
     try {
       this.channel.close().get();
+
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1040]: nit: extra line?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 11]
```diff
@@ -1065,6 +1068,24 @@ public String getChannelName() {
     return this.channel.getFullyQualifiedName();
   }
 
+  // ------ SETTERS ------ //
+  private void setOffsetPersistedInSnowflake(long offsetPersistedInSnowflake) {
+    this.offsetPersistedInSnowflake.set(offsetPersistedInSnowflake);
+    this.snowflakeTelemetryChannelStatus.setOffsetPersistedInSnowflake(offsetPersistedInSnowflake);
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1074]: Instead of getting the offset from here, why we decide to let the telemetry service maintain a copy of all three offsets?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+        this.metricsJmxReporter.getMetricRegistry().getMetrics().keySet().toString());
+    this.metricsJmxReporter.removeMetricsFromRegistry(this.channelName);
+
+    MetricRegistry currentMetricRegistry = this.metricsJmxReporter.getMetricRegistry();
+
+    try {
+      // offsets
+      currentMetricRegistry.register(
+          constructMetricName(
+              this.channelName,
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:89]: Not familiar with JMX, but can a channel name be uniquely identify a metric? What if we have two channels with the same name but different client sequencer?

**Review by sfc-gh-tzhang — COMMENTED** [review]
left some comments, PTAL, thanks!

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+import com.snowflake.kafka.connector.internal.KCLogger;
+import com.snowflake.kafka.connector.internal.metrics.MetricsJmxReporter;
+import com.snowflake.kafka.connector.internal.metrics.MetricsUtil;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class SnowflakeTelemetryChannelStatus {
+  private static final KCLogger LOGGER =
+      new KCLogger(SnowflakeTelemetryChannelStatus.class.toString());
+
+  public static final long NUM_METRICS = 3; // update when new metrics are added
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:33]: Why is this needed? I only see it being used in test, I don't think people will remember update this every time...

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+  private static final long INITIAL_OFFSET = -1;
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:34]: I believe we have a constant for this already

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java (hunk 12 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:89]: Channel names are meant to be[ unique identifiers](https://github.com/snowflakedb/snowflake-kafka-connector/blob/3879f8a3c2349b1ab3955c0bb22e601e08ba35da/src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java#L507) for that particular topic and partition. 

From what I understand, client sequencers are bumped up when a channel is reopened, which means the registry should replace the previous channel name's metrics. This should be ok behavior, wdyt?

Another potential issue is two connectors listening to the same topic, however this is not an issue because the MetricRegistry is created off the unique connector name.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java (hunk 13 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:33]: I think this is important to keep to ensure the IT test using the parameter verifies each new metric added. [SnowflakeTelemetryPipeStatusMetricsIT](https://github.com/snowflakedb/snowflake-kafka-connector/blob/master/src/test/java/com/snowflake/kafka/connector/internal/SnowflakeTelemetryPipeStatusMetricsIT.java#L71) does something similar except it hard codes 14 instead of using a variable, which I think is more confusing

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java (hunk 11 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java]: you're right, the SnowflakeTelemetryPipeStatus creates a copy, however that is not strictly necessary, so i've passed the offsets in TopicPartitionChannel creation

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java (hunk 14 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java]: removed

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java [hunk 15]
```diff
@@ -97,14 +95,15 @@ public void setupEachTest() {
...
+    mockSnowflakeConnectionService = Mockito.mock(SnowflakeConnectionService.class);
     mockTelemetryService = Mockito.mock(SnowflakeTelemetryService.class);
     Mockito.when(mockStreamingClient.isClosed()).thenReturn(false);
     Mockito.when(mockStreamingClient.openChannel(ArgumentMatchers.any(OpenChannelRequest.class)))
         .thenReturn(mockStreamingChannel);
     Mockito.when(mockStreamingChannel.getFullyQualifiedName()).thenReturn(TEST_CHANNEL_NAME);
     this.topicPartition = new TopicPartition(TOPIC, PARTITION);
     this.sfConnectorConfig = TestUtils.getConfig();
-    this.streamingBufferThreshold = new StreamingBufferThreshold(10, 10_000, 1);
+    this.streamingBufferThreshold = new StreamingBufferThreshold(1, 10_000, 1);
```
Comment (sfc-gh-rcheng) [inline src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java:106]: reduced the time flush to 1 second to replicate our recommended settings

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java (hunk 8 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java]: removed, register the guages with TopicPartitionChannel offsets directly instead

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java (hunk 12 above)
Comment (sfc-gh-tzhang) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:92]: what about during rebalance? Any concern about emit metrics from two nodes?

**Review by sfc-gh-tzhang — APPROVED** [review]
LGTM, ship it!

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+import static com.snowflake.kafka.connector.internal.metrics.MetricsUtil.constructMetricName;
+
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.MetricRegistry;
+import com.snowflake.kafka.connector.internal.KCLogger;
+import com.snowflake.kafka.connector.internal.metrics.MetricsJmxReporter;
+import com.snowflake.kafka.connector.internal.metrics.MetricsUtil;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class SnowflakeTelemetryChannelStatus {
```
[the remark this answers is not in the document]
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:29]: will be added in the next PR

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java (hunk 12 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/telemetry/SnowflakeTelemetryChannelStatus.java:92]: There shouldn't be rebalance issues because JmxReporters are created at the serviceSink level. When registering new metrics for a TopicPartitionChannel during rebalance, we remove all existing metrics and then register new ones.
